### ISSUE TITLE
ALC live map loading needs optimization

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#10569]
## Bug description 

The visible area of the live map is not completely populated with adventure lab caches, itseems not optimized to cover the whole viewport in an adaptive way (just guessing)

### Steps to reproduce the problem

- Set global cache type filter to ALC (for easier view of the problem)
- Open live map and zoom out some steps

### Actual result after these steps 

ALCs are only shown in some central part of the map.
If you move the map another cluster around the center of the map gets shown.
Sometimes even some ALC vanish when new ones are loaded after slightly moving the map

### Expected result after these steps 

The map should show all available ALC in the shown viewport.
(Remark: Some limits might apply on low zoom as our internal cache cache / buffer has a certain maximum AFAIK)
So lets say:
The result should be comparable to what we do for "normal" caches on the live map. If the user sees some "normal caches" in an area of the map he assumes, that this is all. But in fact some ALC might be missing.

## c:geo version 

2021.05.08-NB

## Reproducible 

Yes

## Additional context 

IMHO this should not be fixed in a hurry, but carefully compared to what we do with GC:
We want to avoid to many server requests but still assure that we show reliable information...thats sometimes difficult. 
@moving-bits 
You optimized quite a lot for the live map in the past. Maybe you know how to make ALC behave in a comparable way.

**Comment by jakedot** [comment]
Isn't this behaviour deliberate? It's what happens for the other online GC caches too isn't it?

**Comment by Lineflyer** [comment]
> Isn't this behaviour deliberate? It's what happens for the other online GC caches too isn't it?

If you zoom out to much it will also happen to GC and others due to the limit of 1000 caches in total on the viewport (AFAIR).

But the case here is different:
I guess this is connected to the "search by center" instead of "Search in Box" as described by @baiti in another issue.
If you search a center and provide a fixed radius the chance is high, that we do not fill the screen (or at least fill until the internal maximum is reached).
We should at least calculate from the current viewport an indivudual radius to cover this viewport.

**Comment by baiti** [comment]
If a rectangular area needs to be filled, it doesn't help to call a method called `searchByCenter()`, however, that is the **only** search function the ALC API supports.

I have constructed an interface to map a `seaarchByBBox()` to a `searchByCenter()` but the `searchByBBox()` gets never called.

The only option I see, if you want to fill a rectangle is to specify the radius of the "outer circle". Which I would need to calculate first of course. Plus, there is currently no interface to specify the radius as it is hard coded. It seems I have some fundamental lack of understanding hoe a bounding box is filled by the GC connector.

**Comment by baiti** [comment]
In simpler terms: call the LC connectors `searchByBBox()` and I will find a way to fill the outer circle of the given rectangle.

**Comment by baiti** [comment]
I believe I got it now ... it is called `searchByViewport()` ... which is mapped to `searchByBBox()` in the LC connector. So it gets called indeed ;) .... I guess I will do some Haversine math to calculate the length of the diagonal of the bounding box and use half of that length as the radius ...... stay tuned ;)

**Comment by Lineflyer** [comment]
I have in my mind, that the live map is able to support a max. of 1000 caches (in sum of all connectors) before some are dropped or other strange things happen (This is for sure to be optimized...its our famous "cache cache"). 

As long as users operate in "normal" zoom areas (by looking at an area of some kilometers) the maximum to take from the LC connector should be irrelevant. This comes into play if users zoom out a lot....I guess we should stay well below 1000 to cover such cases to avoid LC displacing GC....but to be honest: Someone with more code insides should tell how the live map digests the data from different connectors.

**Comment by baiti** [comment]
If I make a PR, should this also be based on release?

**Comment by baiti** [comment]
If I go to Los Angeles area and set the BBox such that the radius ends up with roughly 300km, I'll get 180 ALCs on the map, would that be "ok"?

**Comment by moving-bits** [comment]
> If I make a PR, should this also be based on release?

yes

> If I go to Los Angeles area and set the BBox such that the radius ends up with roughly 300km, I'll get 180 ALCs on the map, would that be "ok"?

sure - I guess in most cases the amount of GC caches found and the amount of LC caches found within a certain radius will somewhat correlate

**Comment by baiti** [comment]
> > If I make a PR, should this also be based on release?
> 
> yes
> 
> > If I go to Los Angeles area and set the BBox such that the radius ends up with roughly 300km, I'll get 180 ALCs on the map, would that be "ok"?
> 
> sure - I guess in most cases the amount of GC caches found and the amount of LC caches found within a certain radius will somewhat correlate

Ok, I'll throw it in such that you can play with it tomorrow.

**Comment by eddiemuc** [comment]
> > The math should be easy:
> > 
> > SquareRoot((hight/2)^2+(width/2)^2)
> > 
> > (Pythagoras - center to corner distance)
> > 
> > And maybe add a "*1.5" to that number so that we have some "cached content"
> > 
> > of course ... if you imply earth is flat. I think the 1.5 is not necessary as the area of the resulting circle is much bigger than the rectangle anyway
> 
> how about:
> ```
> /**
>      * Calculate distance between two points in latitude and longitude taking
>      * into account height difference. If you are not interested in height
>      * difference pass 0.0. Uses Haversine method as its base.
>      *
>      * lat1, lon1 Start point lat2, lon2 End point el1 Start altitude in meters
>      * el2 End altitude in meters
>      * @returns Distance in Meters
>      */
> 
>     @Nullable
>     private static double distance(final double lat1, final double lat2, final double lon1, final double lon2, final double el1, final double el2) {
>         final int radius = 6371; // Radius of the earth
>         final double latDistance = Math.toRadians(lat2 - lat1);
>         final double lonDistance = Math.toRadians(lon2 - lon1);
>         final double a = Math.sin(latDistance / 2) * Math.sin(latDistance / 2)
>             + Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2))
>             * Math.sin(lonDistance / 2) * Math.sin(lonDistance / 2);
>         final double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
>         double distance = radius * c * 1000; // convert to meters
>         final double height = el1 - el2;
>         distance = Math.pow(distance, 2) + Math.pow(height, 2);
>         return Math.sqrt(distance);
>     }
> ```

Hi @baiti there is already a method for calculating distances between geopoints in c:geo which imho should be used. It is located in class `Geopoint` and is called `distanceTo`

**Comment by baiti** [comment]
> Hi @baiti there is already a method for calculating distances between geopoints in c:geo which imho should be used. It is located in class `Geopoint` and is called `distanceTo`

@eddiemuc  .. Yes ... will fix later today. Just enjoying the beautiful spring weather now and worry about pythagoras later ;)

**Comment by baiti** [comment]
@fm-sys and @eddiemuc ... thx for pointing me to the more efficient solution.

BTW: is there a way to purge the cache cache? Once in a while I would like to have that capability to start fresh.

**Comment by fm-sys** [comment]
The most effective way is to simply create a backup with an empty or basic DB and restore it every time you need it. At least that's the way how I do it ;-)
